std/experimental/allocator/building_blocks/aligned_block_list.d

```diff
@@ -0,0 +1,629 @@
+
+    AlignedBlockNode *root;
+    int numNodes;
+    enum maxNodes = 50;
+
+    static if (isShared)
+    SpinLock lock = SpinLock(SpinLock.Contention.brief);
+
+    private void moveToFront(AlignedBlockNode *tmp)
+    {
+        auto localRoot = cast(AlignedBlockNode*) root;
+        if (tmp == localRoot)
+            return;
+
+        if (tmp.prev) tmp.prev.next = tmp.next;
+        if (tmp.next) tmp.next.prev = tmp.prev;
+        if (localRoot) localRoot.prev = tmp;
+        tmp.next = localRoot;
+        tmp.prev = null;
+
+        root = cast(typeof(root)) tmp;
+    }
+
+    private void removeNode(AlignedBlockNode* tmp)
+    {
+        static if (isShared)
+        import core.atomic : atomicOp;
+
+        auto next = tmp.next;
+        if (tmp.prev) tmp.prev.next = tmp.next;
+        if (tmp.next) tmp.next.prev = tmp.prev;
+        assert(parent.deallocate((cast(void*) tmp)[0 .. alignment]));
+
+        if (tmp == cast(AlignedBlockNode*) root)
+            root = cast(typeof(root)) next;
+
+        static if (isShared)
+        {
+            atomicOp!"-="(numNodes, 1);
+        }
+        else
+        {
+            numNodes--;
+        }
+    }
+
+    private bool insertNewNode()
+    {
+        static if (isShared)
+        import core.atomic : atomicOp;
+
+        void[] buf = parent.alignedAllocate(alignment, alignment);
+        if (buf is null)
+            return false;
+
+        auto localRoot = cast(AlignedBlockNode*) root;
+        auto newNode = cast(AlignedBlockNode*) buf;
+        ubyte[] payload = cast(ubyte[]) buf[AlignedBlockNode.sizeof .. $];
+        newNode.bAlloc = Allocator(payload);
+
+        newNode.next = localRoot;
+        newNode.prev = null;
+        if (localRoot)
+            localRoot.prev = newNode;
+        root = cast(typeof(root)) newNode;
+
+        static if (isShared)
+        {
+            atomicOp!"+="(numNodes, 1);
+        }
+        else
+        {
+            numNodes++;
+        }
+
+        return true;
+    }
+
+public:
+    static if (stateSize!ParentAllocator) ParentAllocator parent;
+    else alias parent = ParentAllocator.instance;
+
+    enum ulong alignment = theAlignment;
+
+    static if (hasMember!(ParentAllocator, "owns"))
+    Ternary owns(void[] b)
+    {
+        return parent.owns(b);
+    }
+
+    bool deallocate(void[] b)
+    {
+        static if (isShared)
+        import core.atomic : atomicOp;
+
+        if (b is null)
+            return true;
+
+        enum ulong mask = ~(alignment - 1);
+        // Round buffer to nearest `alignment` multiple to quickly find
+        // the 'parent' 'AlignedBlockNode'
+        ulong ptr = ((cast(ulong) b.ptr) & mask);
+        AlignedBlockNode *node = cast(AlignedBlockNode*) ptr;
+        if (node.bAlloc.deallocate(b))
+        {
+            static if (isShared)
+            {
+                atomicOp!"-="(node.bytesUsed, b.length);
+            }
+            else
+            {
+                node.bytesUsed -= b.length;
```

Comment:
You might want to consider using your own local function of `atomicOp` that either uses the real `atomicOp` or uses direct assignment - depending on `isShared`